File: mobile/lib/models/activity.dart
class Activity {
  final int? id;
  final String? title;
  final String? description;
  final String? type;
  final DateTime? startDate;
  final DateTime? endDate;
  final String? location;
  final int? maxParticipants;
  final int? currentParticipants;
  final int? organizerId;
  final String? organizerName;
  final bool? isActive;
  final bool? hasAvailableSpots;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  Activity({
    this.id,
    this.title,
    this.description,
    this.type,
    this.startDate,
    this.endDate,
    this.location,
    this.maxParticipants,
    this.currentParticipants,
    this.organizerId,
    this.organizerName,
    this.isActive,
    this.hasAvailableSpots,
    this.createdAt,
    this.updatedAt,
  });

  factory Activity.fromJson(Map<String, dynamic> json) {
    return Activity(
      id: json['id'],
      title: json['title'],
      description: json['description'],
      type: json['type'],
      startDate: json['startDate'] != null ? DateTime.parse(json['startDate']) : null,
      endDate: json['endDate'] != null ? DateTime.parse(json['endDate']) : null,
      location: json['location'],
      maxParticipants: json['maxParticipants'],
      currentParticipants: json['currentParticipants'],
      organizerId: json['organizerId'],
      organizerName: json['organizerName'],
      isActive: json['isActive'],
      hasAvailableSpots: json['hasAvailableSpots'],
      createdAt: json['createdAt'] != null ? DateTime.parse(json['createdAt']) : null,
      updatedAt: json['updatedAt'] != null ? DateTime.parse(json['updatedAt']) : null,
    );
  }
}

File: backend/src/main/java/com/isep/model/Activity.java
package com.isep.model;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import java.time.LocalDateTime;
import java.util.HashSet;
import java.util.Set;

@Entity
@Table(name = "activities")
@Data
@NoArgsConstructor
@AllArgsConstructor
@EntityListeners(AuditingEntityListener.class)
public class Activity {
    
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    
    @Column(nullable = false)
    private String title;
    
    @Column(columnDefinition = "TEXT")
    private String description;
    
    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private ActivityType type;
    
    @Column(name = "start_date", nullable = false)
    private LocalDateTime startDate;
    
    @Column(name = "end_date")
    private LocalDateTime endDate;
    
    @Column(name = "location")
    private String location;
    
    @Column(name = "max_participants")
    private Integer maxParticipants;
    
    @Column(name = "current_participants")
    private Integer currentParticipants = 0;
    
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "organizer_id", nullable = false)
    private User organizer;
    
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "school_id", nullable = false)
    private School school;
    
    @OneToMany(mappedBy = "activity", cascade = CascadeType.ALL, orphanRemoval = true)
    private Set<ActivityRegistration> registrations = new HashSet<>();
    
    @Column(name = "is_active")
    private Boolean isActive = true;
    
    @CreatedDate
    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;
    
    @LastModifiedDate
    @Column(name = "updated_at")
    private LocalDateTime updatedAt;
    
    public enum ActivityType {
        COURSE,
        EXAM,
        WORKSHOP,
        CONFERENCE,
        SPORT,
        TRIP,
        OTHER
    }
    
    public Boolean hasAvailableSpots() {
        return maxParticipants == null || currentParticipants < maxParticipants;
    }
}

